src/Symfony/Component/Messenger/Adapter/PhpAmqp/Connection.php

```diff
@@ -0,0 +1,219 @@
+    private $debug;
+    private $amqpFactory;
+
+    /**
+     * @var \AMQPChannel|null
+     */
+    private $amqpChannel;
+
+    /**
+     * @var \AMQPExchange|null
+     */
+    private $amqpExchange;
+
+    /**
+     * @var \AMQPQueue|null
+     */
+    private $amqpQueue;
+
+    public function __construct(array $connectionCredentials, array $exchangeConfiguration, array $queueConfiguration, bool $debug = false, AmqpFactory $amqpFactory = null)
+    {
+        $this->connectionCredentials = $connectionCredentials;
+        $this->debug = $debug;
+        $this->exchangeConfiguration = $exchangeConfiguration;
+        $this->queueConfiguration = $queueConfiguration;
+        $this->amqpFactory = $amqpFactory ?: new AmqpFactory();
+    }
+
+    public static function fromDsn(string $dsn, array $options = array(), bool $debug = false, AmqpFactory $amqpFactory = null): self
+    {
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new \InvalidArgumentException(sprintf('The given AMQP DSN "%s" is invalid.', $dsn));
+        }
+
+        $pathParts = isset($parsedUrl['path']) ? explode('/', trim($parsedUrl['path'], '/')) : array();
+        $amqpOptions = array_replace_recursive(array(
+            'host' => $parsedUrl['host'] ?? 'localhost',
+            'port' => $parsedUrl['port'] ?? 5672,
+            'vhost' => isset($pathParts[0]) ? urldecode($pathParts[0]) : '/',
+            'queue' => array(
+                'name' => $queueName = $pathParts[1] ?? 'messages',
+            ),
+            'exchange' => array(
+                'name' => $queueName,
+            ),
+        ), $options);
+
+        if (isset($parsedUrl['user'])) {
+            $amqpOptions['login'] = $parsedUrl['user'];
+        }
+
+        if (isset($parsedUrl['pass'])) {
+            $amqpOptions['password'] = $parsedUrl['pass'];
+        }
+
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $parsedQuery);
+
+            $amqpOptions = array_replace_recursive($amqpOptions, $parsedQuery);
+        }
+
+        $exchangeOptions = $amqpOptions['exchange'];
+        $queueOptions = $amqpOptions['queue'];
+
+        unset($amqpOptions['queue']);
+        unset($amqpOptions['exchange']);
+
+        return new self($amqpOptions, $exchangeOptions, $queueOptions, $debug, $amqpFactory);
+    }
+
+    /**
+     * @throws \AMQPException
+     */
+    public function publish(string $body, array $headers = array())
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $this->exchange()->publish($body, null, AMQP_NOPARAM, array('headers' => $headers));
+    }
+
+    /**
+     * Waits and gets a message from the configured queue.
+     *
+     * @throws \AMQPException
+     */
+    public function get(): ?\AMQPEnvelope
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        try {
+            if (false !== $message = $this->queue()->get()) {
```

Comment:
Obviously, there are pros and cons to both methods. Your previous reviews suggested we shouldn't use the `consume` method as it wasn't reliable and had some ["strange behaviour"](https://github.com/symfony/symfony/pull/26632#discussion_r178153477). Let's try to get as close to the fact as possible. Here is a comparison table of the two approaches:

| | `consume` method  | `get` methods |
| -| ------------ | ------------- |
| **Pros**  | <ul><li>Is blocking (low CPU impact for the loop)</li></ul>  | <ul><li>Easy signal handling capabilities</li><li>Will be relatively easy to support `get`ing from multiple queues</li><li>Allow us to adds `ticks` while waiting for the messages</li></ul>
| **Cons**  | <ul><li>Does not handle signals well</li><li>Can't easily handle multiple queues (as blocking)</li><li>Has ["strange behaviour"](https://github.com/symfony/symfony/pull/26632#discussion_r178153477) for ext-amqp < 1.9.3 (which is the last one)</li><li>Messages [might be mixed up](https://github.com/symfony/symfony/pull/26632#discussion_r176930611) between queues</li></ul>  | <ul><li>Higher CPU footprint because of the loop</li><li>It not visible in RabbitMq's UI</li><li>Does not support QoS</li></ul>

Reply:
There is no any strangeness in `basic.conume`, it comes from the way it was used and there is a solution for it a buffer.

I am all for basic consume properly applied .

Reply:
That said, I suggest we go for the `get` method option which seems to be the one that will provide the greater stability. Later, if we some use-cases/issues and believe this was not the correct choice, we can add an option to use the `consume` method instead easily. That way we move forward and stay open to changes :)